src/rules/awaitPromiseRule.ts

```diff
@@ -34,41 +43,43 @@ export class Rule extends Lint.Rules.TypedRule {
 
     public static FAILURE_STRING = "'await' of non-Promise.";
 
-    public applyWithProgram(srcFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
-        return this.applyWithWalker(new Walker(srcFile, this.getOptions(), program));
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        const promiseTypes = new Set(["Promise", ...this.ruleArguments.map(String)]);
+        const tc = program.getTypeChecker();
+        return this.applyWithFunction(sourceFile, (ctx) => walk(ctx, tc, promiseTypes));
     }
 }
 
-class Walker extends Lint.ProgramAwareRuleWalker {
-    public visitNode(node: ts.Node) {
-        if (node.kind === ts.SyntaxKind.AwaitExpression &&
-            !couldBePromise(this.getTypeChecker().getTypeAtLocation((node as ts.AwaitExpression).expression))) {
-            this.addFailureAtNode(node, Rule.FAILURE_STRING);
-        }
+function walk(ctx: Lint.WalkContext<void>, tc: ts.TypeChecker, promiseTypes: Set<string>) {
+    return ts.forEachChild(ctx.sourceFile, cb);
 
-        super.visitNode(node);
+    function cb(node: ts.Node): void {
+        if (isAwaitExpression(node) && !couldBePromise(tc.getTypeAtLocation(node.expression))) {
+            ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
+        }
+        return ts.forEachChild(node, cb);
     }
-}
 
-function couldBePromise(type: ts.Type): boolean {
-    if (Lint.isTypeFlagSet(type, ts.TypeFlags.Any) || isPromiseType(type)) {
-        return true;
-    }
+    function couldBePromise(type: ts.Type): boolean {
+        if (Lint.isTypeFlagSet(type, ts.TypeFlags.Any) || isPromiseType(type)) {
+            return true;
+        }
 
-    if (isUnionType(type)) {
-        return type.types.some(isPromiseType);
-    }
+        if (isUnionType(type)) {
+            return type.types.some(couldBePromise);
+        }
 
-    const bases = type.getBaseTypes();
-    return bases !== undefined && bases.some(couldBePromise);
-}
+        const bases = type.getBaseTypes();
+        return bases !== undefined && bases.some(couldBePromise);
+    }
 
-function isPromiseType(type: ts.Type): boolean {
-    const { target } = type as ts.TypeReference;
-    const symbol = target && target.symbol;
-    return !!symbol && symbol.name === "Promise";
-}
+    function isPromiseType(type: ts.Type): boolean {
+        const { target } = type as ts.TypeReference;
+        const symbol = target && target.symbol;
+        return symbol !== undefined && promiseTypes.has(symbol.name);
+    }
 
-function isUnionType(type: ts.Type): type is ts.UnionType {
-    return Lint.isTypeFlagSet(type, ts.TypeFlags.Union);
+    function isUnionType(type: ts.Type): type is ts.UnionType {
```

Comment:
You don't need to encapsulate that. The encapsulation you get from the file beeing an ES6 module is enough. In fact you should try to keep closures as small as possible to avoid unnecessary work at runtime.

However, it's not that important for now. I'll add this function among others to the next version of `tsutils`. So this one will be replaced anyway.